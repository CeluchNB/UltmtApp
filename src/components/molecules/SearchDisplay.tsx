import React from 'react'
import SearchBar from '../atoms/SearchBar'
import { useColors } from '../../hooks/useColors'
import { useLazyData } from '../../hooks'
import { ActivityIndicator, FlatList, ListRenderItem, View } from 'react-native'

interface SearchDisplayProps<T> {
    value?: string
    search: (q: string) => Promise<T[]>
    renderItem: ListRenderItem<T>
    onChangeText?: (text: string) => void
}

const SearchDisplay = <R,>(props: SearchDisplayProps<R>) => {
    const { value, search, renderItem, onChangeText } = props
    const { colors } = useColors()
    const { data, loading, fetch } = useLazyData(search)

    const onSearch = (q: string) => {
        if (onChangeText) onChangeText(q)
        fetch(q)
    }

    return (
        <View>
            <SearchBar
                placeholder="Search teams..."
                onChangeText={onSearch}
                value={value}
                width="100%"
            />
            {loading && <ActivityIndicator color={colors.textPrimary} />}
            {data && <FlatList data={data} renderItem={renderItem} />}
        </View>
    )
}

export default SearchDisplay
