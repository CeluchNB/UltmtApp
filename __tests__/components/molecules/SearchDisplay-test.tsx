import React from 'react'
import SearchDisplay from '../../../src/components/molecules/SearchDisplay'
import { Text } from 'react-native'
import { QueryClient, QueryClientProvider } from 'react-query'
import {
    fireEvent,
    render,
    screen,
    waitForElementToBeRemoved,
} from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

const client = new QueryClient()

describe('SearchDisplay', () => {
    beforeAll(() => {
        jest.useFakeTimers()
    })

    afterAll(() => {
        jest.useRealTimers()
    })

    it('should render', async () => {
        render(
            <QueryClientProvider client={client}>
                <SearchDisplay
                    placeholder="Search teams..."
                    value="test"
                    search={jest.fn()}
                    renderItem={jest.fn()}
                    onChangeText={jest.fn()}
                />
            </QueryClientProvider>,
        )

        await waitForElementToBeRemoved(() =>
            screen.getByTestId('search-indicator'),
        )
        expect(screen.getByPlaceholderText('Search teams...')).toBeTruthy()
    })

    it('should handle search', async () => {
        const searchFn = jest
            .fn()
            .mockReturnValue(Promise.resolve(['Game 1', 'Game 2']))
        const onChangeFn = jest.fn()
        const renderFn = jest
            .fn()
            .mockImplementation(({ item }) => <Text>{item}</Text>)

        render(
            <QueryClientProvider client={client}>
                <SearchDisplay
                    placeholder="Search teams..."
                    search={searchFn}
                    renderItem={renderFn}
                    onChangeText={onChangeFn}
                />
            </QueryClientProvider>,
        )

        const input = screen.getByPlaceholderText('Search teams...')
        fireEvent.changeText(input, 'test')

        expect(searchFn).toHaveBeenCalledWith('test')
        expect(onChangeFn).toHaveBeenCalledWith('test')

        await waitForElementToBeRemoved(() =>
            screen.getByTestId('search-indicator'),
        )
        expect(renderFn).toHaveBeenCalled()

        expect(screen.getByText('Game 1')).not.toBeNull()
        expect(screen.getByText('Game 2')).not.toBeNull()
    })
})
