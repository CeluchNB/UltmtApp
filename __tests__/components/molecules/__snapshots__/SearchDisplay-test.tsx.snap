// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View>
  <View
    style={
      {
        "alignSelf": "center",
        "display": "flex",
        "flexDirection": "row",
        "width": "100%",
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgba(231, 224, 236, 1)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          {
            "backgroundColor": "#303030",
            "color": "#ffd600",
            "flex": 1,
            "marginBottom": 5,
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#ffd600",
            "bottom": 0,
            "height": 1,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": [
              {
                "scaleY": 0.5,
              },
            ],
            "zIndex": 1,
          }
        }
        testID="text-input-underline"
      />
      <View
        style={
          [
            {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "minHeight": 56,
            },
          ]
        }
      >
        <TextInput
          editable={true}
          focusable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onPressOut={[Function]}
          placeholder="Search teams..."
          placeholderTextColor="#ababab"
          selectionColor="#ffd600"
          style={
            [
              {
                "margin": 0,
              },
              {
                "paddingLeft": 16,
                "paddingRight": 16,
              },
              {
                "height": 56,
              },
              {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              {
                "color": "#ffd600",
                "fontFamily": "System",
                "fontSize": 41.0958904109589,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": undefined,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
              undefined,
            ]
          }
          testID="text-input-flat"
          underlineColorAndroid="transparent"
          value="test"
        />
      </View>
    </View>
  </View>
</View>
`;
